src/client/browserContext.ts

```diff
@@ -96,6 +100,19 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel,
       page._opener.emit(Events.Page.Popup, page);
   }
 
+  private _onRequestFailed(request: network.Request, responseEndTiming: number, failureText: string | undefined) {
```

Comment:
Now looking at these methods, it seems like we should do the logic in them just once (not once in Page and once in BrowserContext).

That means we should remove the corresponding events from the Page interface (in protocol.yml), add the `page: Page?` parameter to them, and emit events here both on the context and the page if any.

To pass the right `Page` in `BrowserContextDispatcher`, we can reach for it as `request.frame().page()`. However, that page might not be initialized yet, so we should only send it once it is initialized, and `reportAsNew` method could mark the page as "initialized".

Sorry that this turns out to be much more complicated than anticipated. If you don't feel like working too much on this, I can pick it up 😄

Reply:
Ah I see, i have cleaned up the code a bit, I hope I did the right thing as per your suggestion, would like you to take a second pass again.